consensus/bor/bor.go

```diff
@@ -976,180 +1029,124 @@ func (c *Bor) APIs(chain consensus.ChainHeaderReader) []rpc.API {
-
-	// method
-	method := "getCurrentSpan"
-	data, err := c.validatorSetABI.Pack(method)
-	if err != nil {
-		log.Error("Unable to pack tx for getCurrentSpan", "err", err)
-		return nil, err
-	}
-
-	result, err := syscall(libcommon.HexToAddress(c.config.ValidatorContract), data)
-	if err != nil {
-		return nil, err
-	}
-
-	// span result
-	ret := new(struct {
-		Number     *big.Int
-		StartBlock *big.Int
-		EndBlock   *big.Int
-	})
-	if err := c.validatorSetABI.UnpackIntoInterface(ret, method, result); err != nil {
-		return nil, err
-	}
-
-	// create new span
-	span := Span{
-		ID:         ret.Number.Uint64(),
-		StartBlock: ret.StartBlock.Uint64(),
-		EndBlock:   ret.EndBlock.Uint64(),
-	}
-
-	return &span, nil
-}
-
-// GetCurrentValidators get current validators
-func (c *Bor) GetCurrentValidators(blockNumber uint64) ([]*Validator, error) {
-	// Use signer as validator in case of bor devent
-	if c.chainConfig.ChainName == networkname.BorDevnetChainName {
-		validators := []*Validator{
-			{
-				ID:               1,
-				Address:          c.signer,
-				VotingPower:      1000,
-				ProposerPriority: 1,
-			},
+	c.closeOnce.Do(func() {
+		if c.HeimdallClient != nil {
+			c.HeimdallClient.Close()
 		}
+	})
 
-		return validators, nil
-	}
-
-	span, err := c.getSpanForBlock(blockNumber)
-	if err != nil {
-		return nil, err
-	}
-	return span.ValidatorSet.Validators, nil
+	return nil
 }
 
 func (c *Bor) checkAndCommitSpan(
 	state *state.IntraBlockState,
 	header *types.Header,
-	chain chainContext,
+	chain statefull.ChainContext,
 	syscall consensus.SystemCall,
 ) error {
 	headerNumber := header.Number.Uint64()
-	span, err := c.GetCurrentSpan(header, state, chain, syscall)
+
+	span, err := c.spanner.GetCurrentSpan(syscall)
 	if err != nil {
 		return err
 	}
+
 	if c.needToCommitSpan(span, headerNumber) {
 		err := c.fetchAndCommitSpan(span.ID+1, state, header, chain, syscall)
 		return err
 	}
+
 	return nil
 }
 
-func (c *Bor) needToCommitSpan(span *Span, headerNumber uint64) bool {
+func (c *Bor) needToCommitSpan(currentSpan *span.Span, headerNumber uint64) bool {
 	// if span is nil
-	if span == nil {
+	if currentSpan == nil {
 		return false
 	}
 
 	// check span is not set initially
-	if span.EndBlock == 0 {
+	if currentSpan.EndBlock == 0 {
 		return true
 	}
 
 	// if current block is first block of last sprint in current span
-	if span.EndBlock > c.config.CalculateSprint(headerNumber) && span.EndBlock-c.config.CalculateSprint(headerNumber)+1 == headerNumber {
+	if currentSpan.EndBlock > c.config.CalculateSprint(headerNumber) && currentSpan.EndBlock-c.config.CalculateSprint(headerNumber)+1 == headerNumber {
 		return true
 	}
 
 	return false
 }
 
-func (c *Bor) getSpanForBlock(blockNum uint64) (*HeimdallSpan, error) {
+func (c *Bor) getSpanForBlock(blockNum uint64) (*span.HeimdallSpan, error) {
 	log.Info("Getting span", "for block", blockNum)
```

Comment:
@0xKrishna should we use debug logs here if this is too frequent?

Reply:
It's okay, Once every 16 blocks, Info can be useful.

Reply:
It's Info only as of now, Let's leave that, Will create separate PR for these kind of changes.